feat: Network Manager UI

vinnyhoward opened the pull request:
## **Description**

**Network Enablement UI integration**

Replaces the legacy network filter with the new Network Manager, allowing users to control which networks are enabled in MetaMask. Enabled networks determine asset visibility and dapp connectivity.
**Key features:**
- Users can add and enable networks through the Network Manager
- Custom Networks and Solana show assets only for the selected network
- Default networks (Ethereum, Arbitrum) display assets across all enabled networks
- Dapp network requests automatically switch to and enable the requested network
- New network additions via dapp connections are auto-enabled


## **Related issues**

Fixes: [5083](https://github.com/MetaMask/MetaMask-planning/issues/5083)

## **Manual testing steps**

1. Run application with `MM_REMOVE_GLOBAL_NETWORK_SELECTOR='true'` to see how this feature interacts with the other GNS removal features.
2. The `@metamask/network-enablement-controller` hasn't been published yet. Here is the [PR](https://github.com/MetaMask/core/pull/6028). Because of this you must build the package locally. Pull down `core` and checkout into the branch `chore/network-enablement-controller` and run `yarn build`
3. Run `yarn && yarn watch:clean`
4. Create a new wallet, or import a new wallet
5. Network filter will initially set to one network, select it to enable networks
6. Enabling Custom networks will remove all other selections
7. Enabling a Default network will multiselect/deselect default networks
8. ☝️ Exception here is Solana. Solana is disabled

## **Screenshots/Recordings**

Update so far: 

https://github.com/user-attachments/assets/df822e7b-0c2f-4e2b-a00a-ef7c2004e61e

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

vinnyhoward committed:
feat: network enablement bottom sheet component

vinnyhoward committed:
feat: additional networks can now be added and selected

vinnyhoward committed:
feat: update custom network UI

vinnyhoward committed:
feat: custom networks can now be viewed and added. currently adding the edit/delete modals for certain networks

vinnyhoward committed:
feat: can now edit/delete networks in the network filter. refactored/cleaned up logic

vinnyhoward committed:
refactor: moved CustomNetworkSelector into its own folder

vinnyhoward committed:
feat: delete rpc modal

vinnyhoward committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into feat-5083-asset-filter-ui

vinnyhoward committed:
feat: integrate new network enablement controller

vinnyhoward committed:
refactor: moved network filtering logic to new component and added feature flag

vinnyhoward committed:
feat: default networks multiselect is now tied to network enablement state

vinnyhoward committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into feat-5083-asset-filter-ui

vinnyhoward committed:
feat: assets can now be sorted based on network enablement state, refactored logic into its own hooks,and updated styles

vinnyhoward committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into feat-5083-asset-filter-ui

vinnyhoward committed:
feat: defi and nft assets are now being filtered

vinnyhoward committed:
refactor: improved performance and readability

vinnyhoward committed:
fix: transactions are now filtered by enabled networks, selecting a default network will deselect custom networks, and added some small minor fixes

vinnyhoward committed:
fix: added network filtering to RPC events, fixed selection logic between popular/custom networks, and fixed tests with feature flag off

vinnyhoward committed:
fix: fixed BridgeView test

vinnyhoward committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into feat-5083-asset-filter-ui

vinnyhoward committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into feat-5083-asset-filter-ui

vinnyhoward committed:
fix: fix styles

vinnyhoward committed:
fix: temp. added the network enablement controller into root for now

vinnyhoward committed:
refactor: moved new hooks into their own folders

vinnyhoward committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into feat-5083-asset-filter-ui

vinnyhoward committed:
Merge branch 'main' into feat-5083-asset-filter-ui

vinnyhoward committed:
Merge branch 'main' into feat-5083-asset-filter-ui

vinnyhoward committed:
fix: remove temp. controller folder

vinnyhoward committed:
test: increased test coverage as well as new tests for new files

vinnyhoward committed:
Merge branch 'feat-5083-asset-filter-ui' of github.com:MetaMask/metamask-mobile into feat-5083-asset-filter-ui

vinnyhoward committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into feat-5083-asset-filter-ui

vinnyhoward committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into feat-5083-asset-filter-ui

vinnyhoward committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into feat-5083-asset-filter-ui

georgewrmarshall reviewed (commented):
Hey @vinnyhoward, just commenting as we get slack notifications for code owner review on open PRs. I'm not sure of the status of this PR but if it's still being worked on can it be changed to draft? If it's ready for review can the `CellMultiSelectWithMenu`, `ListItemMultiSelectWithMenu` and `Cell` updates be broken out into their own PR?

vinnyhoward committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into feat-5083-asset-filter-ui

vinnyhoward committed:
fix: refactor tests that had deprecated methods for network enablement controller

vinnyhoward committed:
fix: minor fixes

vinnyhoward committed:
fix: fixed eslint and ts type issues

vinnyhoward committed:
Merge branch 'main' into feat-5083-asset-filter-ui

vinnyhoward committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into feat-5083-asset-filter-ui

vinnyhoward committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into feat-5083-asset-filter-ui

vinnyhoward committed:
Merge branch 'main' into feat-5083-asset-filter-ui

vinnyhoward committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into feat-5083-asset-filter-ui

vinnyhoward committed:
fix: refactored hook to disallow disabling last network (#18204)

## **Description**

Refactored hook to disallow disabling last network

## **Changelog**

CHANGELOG entry: null

## **Related issues**

Fixes: [18170](https://github.com/MetaMask/metamask-mobile/issues/18170)

## **Manual testing steps**

```gherkin
Feature: Network Manager
  Scenario: user clicks on network to disable
    Given all networks have been disabled except one
    Then nothing should happen and one network will always be enabled
```

## **Screenshots/Recordings**

`~`

### **Before**

`~`

### **After**

`~`

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

Prithpal-Sooriya committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into feat-5083-asset-filter-ui

Prithpal-Sooriya committed:
refactor: fix types

Prithpal-Sooriya commented:
Code coverages is currently at 74.2. A majority of these changes are new changes.
I think we can try bumping this up to 80% but would require some extensive work.

I'm open to having this as a fast follow through.

Prithpal-Sooriya committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into feat-5083-asset-filter-ui

Prithpal-Sooriya committed:
revert: remove small file change

this change isn't strictly necessary, and we can reduce the length of the PR by removing it

Prithpal-Sooriya committed:
refactor: minimise bridging and ramp CO changes

these changes are unnecessary in this large PR

Prithpal-Sooriya committed:
refactor: remove wallet api platform engineer code change

this change I don't think is needed

Prithpal-Sooriya commented on app/components/Views/TransactionsView/index.js:
```diff
@@ -122,14 +128,25 @@ const TransactionsView = ({
-        : allTransactions.filter((tx) => tx.chainId === chainId);
+      let allTransactionsFiltered;
+      if (isRemoveGlobalNetworkSelectorEnabled()) {
+        // TODO: Make sure to come back and check on how Solana transactions are handled
```
Is this TODO still needed?

Prithpal-Sooriya commented on app/constants/network.js:
```diff
@@ -1,4 +1,6 @@
 import { NetworkType, toHex } from '@metamask/controller-utils';
+import { CHAIN_IDS } from '@metamask/transaction-controller';
+import { PopularList } from '../util/networks/customNetworks';
```
changes not needed - removing.

Prithpal-Sooriya committed:
refactor: file change cleanup

these files do not need to be modified in this PR

salimtb reviewed (approved):
lgtm
